Src/Support/Google.Apis.Auth/OAuth2/AccessTokenWithHeaders.cs

```diff
@@ -0,0 +1,66 @@
+﻿/*
+Copyright 2019 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// Represents an access token that can be used to authorize a request.
+    /// The token might be accompanied by extra information that should be sent
+    /// in the form of headers.
+    /// </summary>
+    public class AccessTokenWithHeaders
+    {
+        private const string QuotaProjectHeaderName = "x-goog-user-project";
+
+        private readonly IDictionary<string, IList<string>> _headers = new Dictionary<string, IList<string>>();
+
+        internal AccessTokenWithHeaders(string token)
+        {
+            AccessToken = token;
+            Headers = new ReadOnlyDictionary<string, IList<string>>(_headers);
+        }
+
+        /// <summary>
+        /// An access token that can be used to authorize a request.
+        /// </summary>
+        public string AccessToken { get; }
+
+        /// <summary>
+        /// Extra headers, if any, that should be included in the request.
+        /// </summary>
+        public IReadOnlyDictionary<string, IList<string>> Headers { get; }
+
+        internal AccessTokenWithHeaders WithQuotaProject(string quotaProject)
```

Comment:
I think the `With` prefix is usually used in immutable types, and returns a new instance.
The name `AddQuotaProject` is appropriate here.